staticd/static_vrf.c

```diff
@@ -0,0 +1,196 @@
+/*
+ * STATICd - vrf code
+ * Copyright (C) 2018 Cumulus Networks, Inc.
+ *               Donald Sharp
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the Free
+ * Software Foundation; either version 2 of the License, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; see the file COPYING; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+#include <zebra.h>
+
+#include "vrf.h"
+#include "nexthop.h"
+#include "table.h"
+#include "srcdest_table.h"
+
+#include "static_memory.h"
+#include "static_vrf.h"
+#include "static_routes.h"
+#include "static_vty.h"
+
+static void zebra_stable_node_cleanup(struct route_table *table,
+				      struct route_node *node)
+{
+	struct static_route *si, *next;
+
+	if (node->info)
+		for (si = node->info; si; si = next) {
+			next = si->next;
+			XFREE(MTYPE_STATIC_ROUTE, si);
+		}
+}
+
+static struct static_vrf *static_vrf_alloc(void)
+{
+	struct route_table *table;
+	struct static_vrf *svrf;
+	safi_t safi;
+	afi_t afi;
+
+	svrf = XCALLOC(MTYPE_TMP, sizeof(struct static_vrf));
+
+	for (afi = AFI_IP; afi <= AFI_IP6; afi++) {
+		for (safi = SAFI_UNICAST; safi <= SAFI_MULTICAST; safi++) {
+			if (afi == AFI_IP6)
+				table = srcdest_table_init();
+			else
+				table = route_table_init();
+			table->cleanup = zebra_stable_node_cleanup;
+			svrf->stable[afi][safi] = table;
+		}
+	}
+	return svrf;
+}
+
+static int static_vrf_new(struct vrf *vrf)
+{
+	struct static_vrf *svrf;
+
+	svrf = static_vrf_alloc();
+	vrf->info = svrf;
+	svrf->vrf = vrf;
+
+	return 0;
+}
+
+static int static_vrf_enable(struct vrf *vrf)
+{
+	static_fixup_vrf_ids(vrf->info);
+
+	/*
+	 * We may have static routes that are now possible to
+	 * insert into the appropriate tables
+	 */
+	static_config_install_delayed_routes(vrf->info);
+
+	return 0;
+}
+
+static int static_vrf_disable(struct vrf *vrf)
+{
+	return 0;
```

Comment:
What happens when a VRF is disabled and the ID is reused for another VRF? We're not clearing out the VRF ID that was set in `static_vrf_enable()`, is it possible to get conflicts here or is it OK to keep the IDs cached?

(Also, the answer to this question should be a code comment here because it'll be hard to understand when someone else looks at this in a year...)

Reply:
well it will never be reused( at least most likely will never be reused as that it's a uint32_t number that is increased monotomically per new interface ), I think your point stands though.

What the vrf code does, as I recall, if we shutdown the vrf we disable routes and when it is turned back on we scan the code base and fixup the nhop vrf id's.( at least that is what I recall programming for static routes 4-5 months back ).  All daemons that are vrf aware are probably going to have to be aware of this though.